packages/studio-base/src/players/RosbridgePlayer.ts

```diff
@@ -476,14 +477,129 @@ export default class RosbridgePlayer implements Player {
-      topic.subscribe((message) => {
+
+      this.#rosClient.addTopicCallback(topicName, (message) => {
         if (!this.#providerTopics) {
           return;
         }
         try {
           const buffer = (message as { bytes: ArrayBuffer }).bytes;
           const bytes = new Uint8Array(buffer);
-          const innerMessage = messageReader.readMessage(bytes);
+          // const innerMessage = messageReader.readMessage(bytes);
+          if ( message._topic_type === 'sensor_msgs/msg/LaserScan' ) {
+            let rbounds = message._ranges_uint16.bounds;
+
+            let rdata = this._base64decode(message._ranges_uint16.points);
+
+            let rview = new DataView(rdata);
+            let num_ranges = rdata.byteLength / 2;
+            let points = new Float32Array(num_ranges);
+
+            let rrange = rbounds[1] - rbounds[0];
+            let rmin = rbounds[0];
+
+            for(let i=0; i<num_ranges; i++) {
+              let offset = i * 2;
+
+              let r_uint16 = rview.getUint16(offset, true);
+
+              if(r_uint16 === 65535) {
+                points[i] = NaN;
+                continue; // nan, -inf, inf mapped to 65535
+              }
+
+              let r = r_uint16 / 65534 * rrange + rmin;
+              points[i] = r;
+            }
+            message.ranges = points;
+            message.intensities = points;
+          } 
+          else if ( message._topic_type === 'sensor_msgs/msg/Image' ) {
+            let rdata = message._data_jpeg;
+
+            if ( this.#cachedImage != undefined )
+                message.data = this.#cachedImage;
+
+            // Decode the base64 JPEG to pixel data
+            decodeBase64Jpeg(rdata)
+                .then((pixelData) => {
+                    // Assign the decoded RGB pixel data to message.data
+                    message.data = pixelData; // Uint8Array
+                    this.#cachedImage = pixelData;
+                })
+                .catch((error) => {
+                    console.error('Error decoding image:', error);
+                });
+          }
+          else if ( message._topic_type === 'nav_msgs/msg/OccupancyGrid' ) {
+            let rdata = message._data_jpeg;
+
+            if ( this.#cachedGrid != undefined )
+                message.data = this.#cachedGrid;
+
+            // Decode the base64 JPEG to pixel data
+            decodeBase64Png(rdata)
+                .then((pixelData) => {
+                    // Assign the decoded RGB pixel data to message.data
+                    const decodedA = new Int8Array(pixelData); // Uint8Array
+
+                    const sumsArray = [];
+                    let alen = decodedA.length;
+                    if (decodedA != undefined) {
+                        for (let i=0; i < alen; i+=3) {
+                            sumsArray.push( 0.3 * (decodedA[i] || 0)
+										  + 0.59 * (decodedA[i+1] || 0)
+										  + 0.11 * (decodedA[i+2] || 0) );
```

Comment:
what is this black magic? but a comment about those apparently random numbers

Reply:
I added a comment to clarify this: `          // Transform RGB to grayscale using luminocity method coefficients
`